build: set autoscaling for uiserver

lorenzo-cavazzi opened the pull request:
This should set the autoscaling for the UI server component.
Something is probably wrong because I tried to enable the policy and to load the UI server pod CPU to >50%, but it doesn't spawn a second pod. Maybe @ableuler could help?

fix #1092

/deploy

aledegano commented on helm-chart/renku-ui-server/values.yaml:
```diff
@@ -49,6 +49,13 @@ ingress:
+  enabled: true
+  minReplicas: 1
+  maxReplicas: 5
+  targetCPUUtilizationPercentage: 50
```
Isn't this a bit low as default? Usually scaling out at ~80% is enough to ensure enough capacity will be created before the current running processes are swamped.

lorenzo-cavazzi replied:
Good to know! I'm not very expert on this, I was taking inspiration from this https://github.com/SwissDataScienceCenter/renku-notebooks/blob/master/helm-chart/renku-notebooks/values.yaml#L32-L36
Perhaps that should be changed as well

rokroskar replied:
We set this as the default in the notebooks service for no reason other than it's what is used in the documentation about HPA. It will need to be adjusted, surely.

ableuler commented:
Great that you're adding that from the start. I'd recommend taking a look at the `autoscaling/v2beta2` api. I think it's much nicer and allows you to scale based on memory too.

ableuler commented:
> Something is probably wrong because I tried to enable the policy and to load the UI server pod CPU to >50%, but it doesn't spawn a second pod. Maybe @ableuler could help?

It averages over one minute, see [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/autoscaling/horizontal-pod-autoscaler.md#autoscaling-algorithm) could that be the reason for it not scaling out?

lorenzo-cavazzi commented:
@ableuler I've updated to `autoscaling/v2beta2` and fixed the deployment (the ui-server component wasn't updated correctly).
Not sure if I'm testing in the correct way: I tried to put the cpu under pressure by running `yes > /dev/null &` a few times until I got to >90%, but I still get no more than 1 container in the UI server pod even if I wait a few minutes

![Screenshot_20210414_184703](https://user-images.githubusercontent.com/43481553/114748413-3232c800-9d52-11eb-8e9e-ab9c49d581f0.png)

rokroskar commented:
@lorenzo-cavazzi it might help to see what the horizontal pod autoscaler is doing; you can see its status via `kubectl`, e.g. `kubectl -n renku-ci-ui-1314 get hpa -oyaml`. The utilization you see via `top` might be different than what the metrics server reports - the latter is what is used by the HPA. You can use `kubectl top pod <podname>` to get the numbers from the metrics server.

ableuler commented:
> Something is probably wrong because I tried to enable the policy and to load the UI server pod CPU to >50%, but it doesn't spawn a second pod.

For completeness here: the problem was that there were no resource limits set on the deployment spec. After adding those, the service scales as expected.

lorenzo-cavazzi commented:
Thank you @rokroskar , those commands are very useful!
As mentioned by Andreas, the specific problem was solved by setting the resources (see commit c92e00f132aebbc15480497b865c8eb4c9f32772 ).
Now scaling up (and down) seems to work well

![Screenshot_20210415_093141](https://user-images.githubusercontent.com/43481553/114852830-702bfc80-9de3-11eb-8c63-fff3c337e26a.png)

I ended up setting low resources in the `requests` section, while the `limits` are a bit higher. Let me know if we follow any other best practice.

rokroskar commented:
Ah great! I assumed that might be the problem but when I checked the pod it had limits/requests already :)

ableuler commented on helm-chart/renku-ui-server/values.yaml:
```diff
@@ -49,17 +49,20 @@ ingress:
+resources:
+  limits:
+    cpu: 500m
+    memory: 2Gi
```
Is that realistic? A smaller number here would help the scheduler to place these pods correctly. If you're worried that you might run out of memory, you could enable scaling on memory too.

lorenzo-cavazzi replied:
Like `1Gi` or something closer to the requested amount?

ableuler replied:
Yes, preferably closer to the request. If you're using 5 times the requested amount and memory gets tight on the node, this pod will be a prime candidate for eviction.

lorenzo-cavazzi replied:
Ok I set them down to
```
cpu: 100m
memory: 128Mi
```
